xlators/mgmt/glusterd/src/glusterd-utils.c

```diff
@@ -5467,7 +5467,7 @@ glusterd_compare_friend_data(dict_t *peer_data, dict_t *cmp, int32_t *status,
     }
 
 out:
-    if (ret && arg) {
+    if ((ret || !update) && arg) {
```

Comment:
@Nikhil
   There is a one case still left that needs to be handle.Let's suppose the loop is running for 100 volumes
   for initially 50 volumes there is no issue but for 51 volume the function(glusterd_compare_friend_volume) set the status GLUSTERD_VOL_COMP_RJT  and we are updating ret to 0 so the condition ((ret || !update)) will not be true and memory will not be cleanup.I think either you need to declare a flag or set update to _gf_false in case of goto out from the loop.

Reply:
Yeah, I missed that scenario. Will update it.

Reply:
While the test seems correct now, it needs a bit more of work to understand what it does and when (having needed several updates to the patch is an indication of it).

To make in more clear and clean, why not simply leave the code as it was before and only add `arg = NULL` just after a successful call to `glusterd_launch_synctask()` ?

(btw, looking at the code of `glusterd_launch_synctask()`, I see that it can fail, but the failure is ignored (only a log message is added). Why it doesn't return an error ?!?!?!?)

Reply:
If I understand correctly, `glusterd_launch_synctask` is an async operation so we don't wait for its output. And, if we set `arg` to NULL just after the call to `glusterd_launch_synctask` that would cause the function to fail, as it would still be using that variable.

Reply:
> If I understand correctly, `glusterd_launch_synctask` is an async operation so we don't wait for its output.

`glusterd_launch_synctask()` tries to start an asynchronous task. Once started, it will work independently and asynchronously from the current thread, but it could still fail during setup of the asynchronous request, and this is synchronous. If setup fails, there won't be anyone releasing the memory used by `arg`.

> And, if we set `arg` to NULL just after the call to `glusterd_launch_synctask` that would cause the function to fail, as it would still be using that variable.

`arg` is a local variable. There's no way that setting it to `NULL` will have any impact in an already called function (as long as we don't free it).

Reply:
I agree with xavi we should check the return status of the function glusterd_launch_synctask everywhere, if it returns error we need to cleanup the arg passed to the function.

Reply:
Ack. I will update the references of the function to check for the return value.

Reply:
One thing though, there are few calls where we pass NULL for the `*opaque` arg. I think we can ignore the check for a return value in those cases?

Reply:
In those cases we don't need to release anything, but the caller still needs to know if the asynchronous operation has been started or not to decide what to do (otherwise the error may be silently ignored and the user may assume that whatever she tried it succeeded).

However I haven't checked how many times this functions is called and in what contexts. If the changes become too big, we can fix memory leaks here and leave the remaining changes for another patch.

Reply:
Ok. Luckily, the change is not big but it is used in some critical places like the glusterd initialization(init() in glusterd.c) and glusterd friend updates (glusterd_friend_sm in glusterd-sm.c).

And, both of the above functions don't need to free anything in case of error. I can add a check for `ret` value and exit in case of error, but honestly, I am not sure if it will not cause any wider impact in the working of glusterd.

Reply:
The error should never happen, so there shouldn't be any difference. But in the rare case it could happen, current code simply logs a critical message and continues normally (probably it won't do what it should though). With the change, the error will be propagated, so something will ultimately fail because of this problem and the user will be notified. I think it should be fine. I'll review the changes with care anyway.

Reply:
Ack. I will update the code accordingly.

Reply:
Done.